Replaced ACK_FAILURE with RESET

technige opened the pull request.

technige committed:
Replaced ACK_FAILURE with RESET

technige committed:
Tweaked altered error messages

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/handlers/ResetResponseHandler.java:
```diff
@@ -56,8 +56,11 @@ public final void onRecord( Value[] fields )
-        messageDispatcher.unMuteAckFailure();
-        resetCompleted( completionFuture, success );
+        messageDispatcher.clearCurrentError();
+        if (completionFuture != null)
```
Why this null check is added? Did we changed as old behaviour here? 
Why this future could be null and is it safe to not have an `else` here?

zhenlineo reviewed (commented):
It is really happy to see `muteAckFailure` removed!
I got one question after reading this PR. 
Finally you might notice that your formatting is a bit different from the existing code.

lutovich commented on driver/src/test/java/org/neo4j/driver/v1/integration/TransactionAsyncIT.java:
```diff
@@ -1157,7 +1157,7 @@ public void shouldFailToCommitWhenRunFailureIsConsumed()
         catch ( ClientException e )
         {
-            assertThat( e.getMessage(), startsWith( "Transaction rolled back" ) );
+            assertThat( e.getMessage(), startsWith( "No current transaction to commit." ) );
```
This message change is quite weird. It exposes an implementation detail that driver resets a connection after an error. Do you think it is okay?

lutovich commented on driver/src/main/java/org/neo4j/driver/internal/async/inbound/InboundMessageDispatcher.java:
```diff
@@ -140,13 +132,14 @@ public void handleFailureMessage( String code, String message )
-        // try to write ACK_FAILURE before notifying the next response handler
-        ackFailureIfNeeded();
+        // write a RESET to "acknowledge" the failure
+        queue( new ResetResponseHandler( this, null ) );
```
Maybe better to create a constructor that only takes an `InboundMessageDispatcher` and uses `null` inside `ResetResponseHandler`. Just to not pass `null` explicitly as the second parameter.

lutovich reviewed (commented):
@nigelsmall review completed. Made two comments. It would also be nice to add a test that asserts valid errors are thrown in an explicit transaction after a failed run. Smth like this for blocking and async API:

```java
Transaction tx = session.beginTransaction();
        
tx.run( "RETURN invalid" ).consume(); // should throw syntax error
tx.run( "RETURN 1" ).consume(); // should throw some meaningful error
        
tx.close();
```

lutovich committed:
Small cleanup in ResetResponseHandler

lutovich committed:
Treat transaction as rolled back after any failure

Every failure is now "acknowledged" by sending a RESET message. It
makes database cleanup the state of the transaction and forget about
it. Thus every failure is fatal and transaction can't execute more
queries or commit.

This commit makes every failure treated as fatal for transactions.

lutovich committed:
Wait for all errors to arrive before closing tx

Results of all queries need to be consumed before transaction is
committed or rolled back. This is done by `ResultCursorsHolder`
which keeps all cursors and queries them for errors. Previously,
it only waited for the first failure to arrive and there might've
been some more queries receiving results.

This commit fixes the problem by making the `ResultCursorsHolder`
properly await for all result cursors to be consumed. It does not do
blocking await but structures the result future to only be completed
when all results have arrived.

lutovich committed:
Add couple ITs for errors in explicit transactions

lutovich committed:
Adjust error message assertions
